tests/xmap_test.py

```diff
@@ -272,11 +276,204 @@ def tearDown(self):
 
 
+LHSSpec = RHSSpec = Tuple[int, ...]
+DimSpec = Tuple[LHSSpec, RHSSpec]
+
+class PdotSpec(NamedTuple):
+  """Specifies positional axes to contract or batch via mapping or position."""
+  map_contract: DimSpec
+  pos_contract: DimSpec
+  map_batch: DimSpec
+  pos_batch: DimSpec
+
+  @property
+  def num_mapped(self):
+    return len(self.lhs_mapped)
+
+  @property
+  def num_mapped_contract(self):
+    return len(self.map_contract[0])
+
+  @property
+  def lhs_mapped(self):
+    return [*self.map_contract[0], *self.map_batch[0]]
+
+  @property
+  def rhs_mapped(self):
+    return [*self.map_contract[1], *self.map_batch[1]]
+
+  @property
+  def pos_contract_after_mapping(self):
+    lhs_pos_contract_ = [i - sum(j < i for j in self.lhs_mapped)
+                         for i in self.pos_contract[0]]
+    rhs_pos_contract_ = [i - sum(j < i for j in self.rhs_mapped)
+                         for i in self.pos_contract[1]]
+    return (lhs_pos_contract_, rhs_pos_contract_)
+
+  @property
+  def pos_batch_after_mapping(self):
+    lhs_pos_batch_ = [i - sum(j < i for j in self.lhs_mapped)
+                      for i in self.pos_batch[0]]
+    rhs_pos_batch_ = [i - sum(j < i for j in self.rhs_mapped)
+                      for i in self.pos_batch[1]]
+    return (lhs_pos_batch_, rhs_pos_batch_)
+
+  @property
+  def all_contract(self):
+    lhs_contract = [*self.map_contract[0], *self.pos_contract[0]]
+    rhs_contract = [*self.map_contract[1], *self.pos_contract[1]]
+    return lhs_contract, rhs_contract
+
+  @property
+  def all_batch(self):
+    lhs_batch = [*self.map_batch[0], *self.pos_batch[0]]
+    rhs_batch = [*self.map_batch[1], *self.pos_batch[1]]
+    return lhs_batch, rhs_batch
+
+def all_pdot_specs(lhs_shape: Tuple[int], rhs_shape: Tuple[int]
+                   ) -> Generator[PdotSpec, None, None]:
+  """Test utility to generate all valid pdot specs for given shapes."""
+  for dim_nums in all_dot_general_dimension_numbers(lhs_shape, rhs_shape):
+    yield from map_some_axes(dim_nums)
+
+def all_dot_general_dimension_numbers(
+    lhs_shape: Sequence[int], rhs_shape: Sequence[int]
+  ) -> Generator[DotDimensionNumbers, None, None]:
+  """Test utility to generate all valid dot_general specs for given shapes."""
+  # The recursive 'helper' function generates all possible DotDimensionNumbers
+  # while excluding certain axis positions. The base case in the first yield is
+  # not to assign any additional axes to be batched or contracted, while the
+  # recursive cases in the latter two yields come from choosing equal-sized axes
+  # to batch or contract (recursing to possibly choose more axes).
+  def helper(excluded1: Set[int], excluded2: Set[int]):
+    yield ((), ()), ((), ())
+    for i, d1 in enumerate(lhs_shape):
+      if i not in excluded1:
+        for j, d2 in enumerate(rhs_shape):
+          if j not in excluded2 and d1 == d2 and j <= i:
+            for contract, batch in helper(excluded1 | {i}, excluded2 | {j}):
```

Comment:
I think that makes the position computations more complicated. That is, when the recursive call returns axis positions, then the caller would need to increment them depending on whether it had eliminated an axis.

I think it's better to keep things 'flat' so as to avoid as much axis incrementing bookkeeping as possible.

WDYT? LMK if that's unclear and I can explain better.

Reply:
On the other hand, doing it this way this was a source of a bug that caused duplicate cases to be generated :)

Reply:
It turned out it wasn't this approach that caused the duplication bug (I tried it recursing on sub-shapes and it was just as hard!). I had just messed up.